Adds an inner fill to the pointer and the ability to stroke inside the pointer

jonathan-livly opened the pull request:
Adds an inner fill and the ability to stroke inside the pointer as seen here
<img width="457" alt="image" src="https://user-images.githubusercontent.com/52213698/169098630-0c63ac9f-1d5e-4ce6-a817-a73572be24a1.png">

This PR is slightly more opinionated with the name changes since it was written as a drop in replacement into our code base since I was unsure if you were still accepting PRs at the time. Deprecated some of the names to make it easier for our team to reason about the code after the updates. 

Let me know if you disagree with any of the changes.

tankery commented:
Hi, @jonathan-livly , Thanks for the PR. I understand that the changed name may sounds more intuitive. But unless there is a strong reason, we'd prefer to keep the existing APIs.

The original design of this name for the "pointerStrokeWidth", is actually based on the entire circle bar perspective. And the name is also matched with other properties such as `circleStrokeWidth` and `pointerHaloWidth`.

Regarding the new property you added, the `pointerFillInnerStrokeWidth`, it's a good design to have something like this. We already already have a similar concept, the `pointerHaloBorderWidth`. Given this, do you think we could naming it as something like `pointerBorderWidth` and `pointerBorderColor`?

Also if we are adding a new width to the pointer, please do consider the calculation of "pointerHaloPaint", which should display outside the boarder when `userIsMovingPointer`.

jonathan-livly commented:
> Hi, @jonathan-livly , Thanks for the PR. I understand that the changed name may sounds more intuitive. But unless there is a strong reason, we'd prefer to keep the existing APIs.
Understandable, I reverted those changes

> The original design of this name for the "pointerStrokeWidth", is actually based on the entire circle bar perspective. And the name is also matched with other properties such as `circleStrokeWidth` and `pointerHaloWidth`.
> 
> Regarding the new property you added, the `pointerFillInnerStrokeWidth`, it's a good design to have something like this. We already already have a similar concept, the `pointerHaloBorderWidth`. Given this, do you think we could naming it as something like `pointerBorderWidth` and `pointerBorderColor`?
> 
> Also if we are adding a new width to the pointer, please do consider the calculation of "pointerHaloPaint", which should display outside the boarder when `userIsMovingPointer`.

So the way that I designed the effect was that there is a another circle drawn over the pointer created by `pointerPaint` with `pointerFillPaint`. It will always be at most as big as pointerFillPaint and `pointerFillInnerStrokeWidth`can only be used to reduce the size. I added a check to make sure negative numbers can't be used after your comment. As it shrinks the `pointerPaint` is effectively stroking `pointerFillPaint` by the size of `pointerFillInnerStrokeWidth`.

I'm not super happy with the names I came up with either, `pointerBorderWidth` kind of makes sense to me and I'd be happy to make that change, but `pointerBorderColor` I think would be confusing based on the way that it works. Maybe `pointerInnerBorderWidth` or `pointerInnerFillBorderWidth` and `pointerInnerFillColor`

tankery commented on README.md:
```diff
@@ -71,6 +71,8 @@ app:cs_start_angle = "float"
 app:cs_disable_progress_glow = "boolean"
 app:cs_hide_progress_when_empty = "boolean"
+"cs_pointer_fill_color" = "color"
+"cs_pointer_fill_inner_stroke_width" = "dimension"
```
Please add the `app:` namespace.

tankery commented on circularSeekBar/src/main/java/me/tankery/lib/circularseekbar/CircularSeekBar.kt:
```diff
@@ -85,11 +85,16 @@ class CircularSeekBar @JvmOverloads constructor(
+    /**
+     * `Paint` instance used to draw the pointer's fill. Will effectively use [pointerPaint] as stroke.
+     */
+    private val pointerFillPaint: Paint = Paint()
```
It looks good to have an smaller circle inside the original pointer. To help user understanding this idea, do you think a name like "pointerInner" make more sense than "pointerFill"?

Which means, you may add two properties:

1. pointerInnerColor
2. pointerInnerStrokeWidth


Or you may also consider something like "pointerOverlay" to reflect the actually implementation is drawing an overlay on top of existing pointer.

jonathan-livly replied:
I actually really like `pointerOverlay` alot. Will make that change

tankery commented on circularSeekBar/src/main/java/me/tankery/lib/circularseekbar/CircularSeekBar.kt:
```diff
@@ -545,9 +575,10 @@ class CircularSeekBar @JvmOverloads constructor(
         pointerStrokeWidth = attrArray.getDimension(
             R.styleable.cs_CircularSeekBar_cs_pointer_stroke_width,
-            DEFAULT_POINTER_STROKE_WIDTH
+            DEFAULT_POINTER_WIDTH
```
Wondering why we change this? `DEFAULT_POINTER_STROKE_WIDTH` sounds more unified with `pointerStrokeWidth`

tankery commented on circularSeekBar/src/main/java/me/tankery/lib/circularseekbar/CircularSeekBar.kt:
```diff
@@ -705,6 +747,12 @@ class CircularSeekBar @JvmOverloads constructor(
 
+        pointerFillPaint.set(pointerPaint)
+        pointerFillPaint.color = pointerFillColor
+        pointerFillPaint.style = Paint.Style.FILL_AND_STROKE
```
This I believe should be `Paint.Style.STROKE`, otherwise the view will have visual issue when the `pointerAngle` > 0.

tankery commented on circularSeekBar/src/main/java/me/tankery/lib/circularseekbar/CircularSeekBar.kt:
```diff
@@ -705,6 +747,12 @@ class CircularSeekBar @JvmOverloads constructor(
+        pointerFillPaint.color = pointerFillColor
+        pointerFillPaint.style = Paint.Style.FILL_AND_STROKE
+        pointerFillPaint.strokeWidth =
+            (pointerStrokeWidth - pointerFillInnerStrokeWidth).coerceAtLeast(0f)
```
If we use a concept like "pointerInnerStrokeWidth" or "pointerOverlayStrokeWidth", we could simply apply the pointerInnerStrokeWidth here, and let the developer to do the math, and give developer the flexibility.

tankery reviewed (commented):
Thanks, I see what you mean, added some inline comments for your changes, please check if it make sense.

tankery commented on circularSeekBar/src/main/java/me/tankery/lib/circularseekbar/CircularSeekBar.kt:
```diff
@@ -31,12 +31,12 @@ import android.content.Context
 import android.graphics.BlurMaskFilter
 import android.graphics.PathMeasure
 import android.graphics.Canvas
+import android.graphics.Paint.Cap
```
nit: Please try keep the imports in order

tankery commented on circularSeekBar/src/main/java/me/tankery/lib/circularseekbar/CircularSeekBar.kt:
```diff
@@ -705,6 +748,12 @@ class CircularSeekBar @JvmOverloads constructor(
+        pointerOverlayPaint.color = pointerOverlayColor
+        pointerOverlayPaint.style = Paint.Style.STROKE
+        pointerOverlayPaint.strokeWidth =
+            (pointerStrokeWidth - pointerOverlayBorderWidth).coerceAtLeast(0f)
```
If we use the concept of "overlay", than the border width of this overlay looks more like it's own width (circle diameter).

So at here we may directly set the `pointerOverlayPaint.strokeWidth = pointerOverlayBorderWidth`

tankery commented on circularSeekBar/src/main/java/me/tankery/lib/circularseekbar/CircularSeekBar.kt:
```diff
@@ -1128,6 +1179,7 @@ class CircularSeekBar @JvmOverloads constructor(
         circleColor = savedState.getInt("circleColor")
         circleProgressColor = savedState.getInt("circleProgressColor")
         pointerColor = savedState.getInt("pointerColor")
+        pointerOverlayColor = savedState.getInt("pointerOverlayColor")
```
Should we also save `pointerOverlayBorderWidth`?

tankery reviewed (commented):
Thanks for the update, looks good overall, just some small questions.

jonathan-livly commented:
Updated. Sorry for being a little slow, our repository is quite a bit ahead so haven't had much time to go back, but we would like to not have to use our fork. 

Updated per the other code comments. I think we are just thinking about the `pointerOverlayBorderWidth` in different mindsets. The technical implementation of it is as an overlay which is why I believe you think it should be sized directly, but from a usage stand point I don't think it really makes to get the goal of its implementation, which is a border for the pointer. 

Ideally the pointer would simply be a circle of a certain width with a border/stroke of a certain width, which is how I believe most designers would be implementing the design specs they are passing off. With that in mind I think its an unnecessary step to make them do the math to determine the size of overlay to get the border they want.

With that in mind I think its a minor difference in how we are thinking so if you feel very strongly, its a change I can make but I think it makes the API harder to use, although only very slightly.

tankery commented:
@jonathan-livly I do appreciate the effort you put to improve this library.

I understand your current approach would be more easy to apply from a design specs. And I agree that the current naming pattern for this library is not good enough for many cases.

But for this specific PR, I would think that naming consistency has a higher priority.

As you would see, we currently already have many properties with names like "**Width", such as `pointerStrokeWidth`, `pointerHaloWidth`, `pointerHaloBorderWidth`, etc. And they all represent the `Paint.strokeWidth` when drawing that object.

So if we are adding a new property with the same naming pattern, but the `Width` is not the same meaning as before, this may require more explanation and context switching for developers.

So I would still recommend just setting the `pointerOverlayBorderWidth` as the `Paint.strokeWidth` for the overlay object.

And after this PR. Feel free to rename properties to a better understanding pattern. But please do consider that this circular seek bar is built not only for "circle/point" pointers, but also for "arc/sector" pointers.
